### ISSUE TITLE
linux proc collectors - these should be migrated to sys

### ISSUE THREAD

**Issue by Stoatwblr** [netdata/netdata#7319]
For various reasons (primarily inconsistency of presentation in /proc), it's generally regarded as a better idea to use /sys as the first port of call when reading data than /proc 

This is particularly relevant when reading anything related to hardware or system parameters - there's virtually nothing in /proc that isn't available in /sys, (usually with greater detail)

Various efforts have been underway to "deprecate" or cleanup /proc for over a decade and as a result this area has mostly stagnated for at least 15 years (one example: If you want scsi_tape stats, you can _ONLY_ get them in /sys)

**Comment by Stoatwblr** [comment]
Some expansion of this subject is at http://landoflinux.com/linux_procfs_sysfs.html and https://dwaves.org/2017/05/29/linux-difference-between-proc-sys-and-dev-sysfs/

effectively: /proc should only be used for mapping into processes - everything else is in /sys

(This philosophy predates netdate and hardware/kernel collectors shouldn't have been operating in /proc in the first place)

**Comment by vlvkobal** [comment]
What about supporting old Linux distributions? For example, CentOS 6. Is `proc` already deprecated for any data?

**Comment by Stoatwblr** [comment]
/sys has been the _standard_ way of representing this stuff since before 2005 (kernel 2.6) If you want to hold up CentOs as an example you're talking about supporting CentOS _4_ and that was last updated in 2009.

CentOS 6 was released in 2011 and fully supported /sys and sysfs from its initial release - for that matter, so did Centos5 and that was released in 2007

**Comment by mfundul** [comment]
Let's look at 2 very important projects, [procps](https://gitlab.com/procps-ng/procps/blob/master/proc/sysinfo.c#L58) and [sysstat](https://github.com/sysstat/sysstat/blob/b0296bccc386ac1f3b283ccabb4e9534cb700ca4/common.h#L71).

Most of those `/proc/` entries contain aggregated information that is really nowhere to be found in `/sys/`. The same applies for some recent pull requests we merged such as `/proc/slabinfo`. For many `/proc/` entries it's either impossible to find the information in `/sys/` or impractical to process and aggregate it when it's already there in `/proc/` provided by the kernel itself with a view that is atomic. Parsing many different `/sys/` files and aggregating the information manually in a non-atomic manner should be avoided.

Moreover, the fact that many linux tools use `/proc/` aggressively also guarantees that there is no risk of deprecation here. I don't think further effort is warranted to refactor netdata from `/proc/` to `/sys/` without any explicitly stated gains.

I believe we should treat this situation on a case by case basis, e.g. at the chart level. For instance, are there any suggestions to improve a specific chart that will gain from additional information from `/sys/`? Are there any suggestions for a completely new chart that is based on `/sys/`?

**Comment by Stoatwblr** [comment]
tapestats for starters. 

/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/dev
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/defined
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/power
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/power/control
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/power/async
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/power/runtime_enabled
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/power/runtime_active_kids
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/power/runtime_active_time
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/power/autosuspend_delay_ms
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/power/runtime_status
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/power/runtime_usage
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/power/runtime_suspended_time
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/stats
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/stats/write_cnt
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/stats/read_ns
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/stats/write_byte_cnt
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/stats/io_ns
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/stats/other_cnt
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/stats/resid_cnt
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/stats/read_byte_cnt
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/stats/in_flight
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/stats/read_cnt
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/stats/write_ns
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/default_density
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/device
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/subsystem
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/uevent
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/default_compression
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/options
/sys/devices/pci0000:00/0000:00:03.0/0000:05:00.0/host11/port-11:0/end_device-11:0/target11:0:0/11:0:0:0/scsi_tape/nst3/default_blksize

 For disks, the level of information available dwarves /proc:

**Comment by Stoatwblr** [comment]
In any case, the stated philosophy of linux going forward from 2005 is that /proc is for process interfacing and /sys is for kernel/device interfacing - if you ignore that you do do at your peril.

There have been no new /proc/ items for a very long time.

/proc/slabinfo is a hangover from very old days. It was there in kernel 0.93 and has been grandfathered in. That doesn't mean it will always be there and various slab items have been removed over the years (the slabstat interface went away in 2.6.12) 
   - again, /sys/kernel/slab contains _far_ more detail than the ancient /proc interface


The one item per file philosophy of Linux /sysfs might seem ridiculous to you, but it is the _STANDARDISED_ way of doing things, so get used to it because it's not going away. If you want to argue about that pick a fight with the kernel devs, but that argument was had and settled in 2001.

**Comment by mfundul** [comment]
I'm not arguing about philosophy at all. I actually agree that `/proc/` is a mess. But that is not the point.

I'm simply saying that some things that work for netdata (and other tools) will not change without a clear benefit, especially if the effort is significant.

**Comment by Stoatwblr** [comment]
The primary benefit is that netdata plugins will continue to work without an update in the future resulting in an "emergency update" being required. 

Devs can either plan ahead, or act like a bunch of headless chickens when it happens.

Relying on grandfathered and deprecated interfaces is a high-risk policy. 

At some point they WILL be switched off. It gets brought up on a semi-regular basis on the kernel developer mailing lists and a lot of this very old interface code relies on one or two people who are retiring from active developing, so when the symbols become incompatible, it's being ripped out rather than updated (This has already happened to several chunks of older chunks of kernel code)

**Comment by mfundul** [comment]
Can you please link a discussion so that we can evaluate the risk?

**Comment by Ferroin** [comment]
@Stoatwblr One of the things you seem to be ignoring here (as are the kernel devs) is that there's a lot of stuff that can be atomically accessed in `/proc` that cannot in `/sys`. The slab stats are a prime example of this, you can get all of them with a _single atomic_ read of `/proc/slabinfo`, but that's impossible through `/sys/kernel/slab`.  Data collection for stuff like this kind of needs to be atomic, otherwise it's significantly less meaningful.

Additionally, by Linus' own standards, none of what we're using in `/proc` is likely to ever go away as multiple tools are actively using them (as long as there's at least one active user and a feature is not blocking implementation of other features, it won't get removed, end of story).

That said, I agree that disk stats should probably be pulled from `/sys` instead of `/proc/diskstats`. There's a lot more info in the `/sys` interface, and it's not as much of an issue if data collection for disks isn't atomic relative to other disks.

--

@mfundul `/proc/acpi` and `/proc/driver/rtc` are concrete examples of solidly deprecated and often no-longer enabled interfaces in `/proc`. Both are still technically in the kernel, but most distros don't enable either at all these days, and they're really only hanging on because of legacy third-party code that makes use of them.

**Comment by Saruspete** [comment]
Hello,

TL;DR: +1 with @Ferroin 

Rule N°1 of the kernel is  **We do not break userspace**

The lifespan of code in the kernel is really very high. They just started to orphan floppy disk drivers on v5.4 ([47d6a76](https://github.com/torvalds/linux/commit/47d6a76)), and only because it exposed VMs to a security risk. 

As stated by Ferroin, the risk may come from distributions that would disable some of these /proc entries in their kernel config, but even that is something unlikely to happen (and could be reverted). About the `/proc/acpi` deprecation [drivers/acpi/Kconfig](https://github.com/torvalds/linux/blob/master/drivers/acpi/Kconfig)) it's only the elements that were implemented in /sys that were removed.
For `/proc/driver/rtc` I still have it (along with nvram) on my Fedora 29 (Kernel 5.2).

The main difference between `/proc` (procfs) and `/sys` (sysfs) is procfs is a call to a kernel function to do access and formatting of a global view (thus, an atomic operation), while sysfs provides a direct access to the variables (even in write access). So that's a lot of syscalls and non-atomic operations that would affect the correctness of the metrics.

Another note for slabinfo: /sys is only available for SLUB-type (not SLAB nor SLOB), while /proc is available for all types of slab.


Finally, nowhere in the official kernel doc did I found a note about this deprecation.

Not a bug.

> Talk is cheap, show me the code -- LT

**Comment by Ferroin** [comment]
> For /proc/driver/rtc I still have it (along with nvram) on my Fedora 29 (Kernel 5.2).

Yeah, a lot of distros still provide it, but essentially nothing uses it anymore except people and a few rare third-party non-FOSS tools, so there are a handful disabling it, and quite a few people who build custom kernels disable it (saves a few kB of kernel size at runtime and reduces potential attack surface).

**Comment by ilyam8** [comment]
From the discussion above i see that there is no need to migrate linux proc collectors to sys.

Still there are some good points by @Ferroin , for example

> That said, I agree that disk stats should probably be pulled from /sys instead of /proc/diskstats. There's a lot more info in the /sys interface, and it's not as much of an issue if data collection for disks isn't atomic relative to other disks.

@vlvkobal take a look, if we need it, please open a specific issue.
